confluent_kafka/src/get_metadata_impl.c

```diff
@@ -0,0 +1,385 @@
+		if (err == -1) {
+			goto error;
+		}
+	}
+
+	return py_partitions;
+
+error:
+	Py_XDECREF(py_partitions);
+	Py_XDECREF(py_part);
+	Py_XDECREF(py_topic);
+	return NULL;
+}
+
+static PyObject *
+topics_to_dict(int topic_cnt, const rd_kafka_metadata_topic_t *topics,
+       	       const struct broker_entry *brokers, int broker_cnt)
+{
+	int idx;
+	PyObject *py_topics = NULL;
+	PyObject *py_partition_list = NULL;
+
+	if ((py_topics = PyDict_New()) == NULL) {
+		goto error;
+	}
+
+	for (idx=0; idx < topic_cnt; ++idx) {
+		py_partition_list = make_partition_list(topics[idx].topic,
+							topics[idx].partition_cnt,
+							topics[idx].partitions,
+						       	brokers, broker_cnt);
+
+		if (py_partition_list == NULL) {
+			goto error;
+		}
+
+		if (PyDict_SetItemString(py_topics, topics[idx].topic, py_partition_list) == -1) {
+			goto error;
+		}
+
+		Py_DECREF(py_partition_list);
+	}
+
+	return py_topics;
+
+error:
+	Py_XDECREF(py_topics);
+	Py_XDECREF(py_partition_list);
+	return NULL;
+}
+
+static struct broker_entry *
+make_broker_objects(int broker_cnt, const rd_kafka_metadata_broker_t *brokers)
+{
+	int idx;
+	PyObject *py_broker = NULL;
+
+	struct broker_entry *py_brokers = calloc(broker_cnt, sizeof(struct broker_entry));
+	if (!py_brokers) {
+		return NULL;
+	}
+
+	for (idx = 0; idx < broker_cnt; ++idx) {
+		if ((py_broker = PyDict_New()) == NULL) {
+			goto error;
+		}
+
+		if (SET_LONG_IN_DICT(py_broker, &brokers[idx], id) == -1) {
+			goto error;
+		}
+
+		if (SET_CHAR_IN_DICT(py_broker, &brokers[idx], host) == -1) {
+			goto error;
+		}
+
+		if (SET_LONG_IN_DICT(py_broker, &brokers[idx], port) == -1) {
+			goto error;
+		}
+
+		py_brokers[idx].id = brokers[idx].id;
+		py_brokers[idx].py_broker = py_broker;
+	}
+
+	return py_brokers;
+
+error:
+	for (idx = 0; idx < broker_cnt; idx++) {
+		Py_XDECREF(py_brokers[idx].py_broker);
+	}
+
+	free(py_brokers);
+
+	return NULL;
+}
+
+static PyObject *
+metadata_to_dict(const rd_kafka_metadata_t *metadata)
+{
+    PyObject *py_topics = NULL;
+    struct broker_entry *py_brokers = NULL;
+    int idx;
+
+    py_brokers = make_broker_objects(metadata->broker_cnt, metadata->brokers);
+    if (py_brokers == NULL) {
+        return NULL;
+    }
+
+    py_topics = topics_to_dict(metadata->topic_cnt, metadata->topics,
+			       py_brokers, metadata->broker_cnt);
+    for (idx = 0; idx < metadata->broker_cnt; idx++) {
+	    Py_DECREF(py_brokers[idx].py_broker);
+    }
+
+    free(py_brokers);
+
+    return py_topics;
+}
+
+
+PyObject *
+Kafka_generic_get_metadata(Handle *self, PyObject *args, PyObject *kwargs)
+{
+    CallState cs;
+    PyObject *result = NULL;
+
+    rd_kafka_resp_err_t err;
+    const rd_kafka_metadata_t *metadata = NULL;
+    rd_kafka_topic_t *only_rkt = NULL;
+
+    int all_topics = 1;
+    const char *topic = NULL;
+    double timeout = -1.0f;
+    static char *kws[] = {"all_topics", "topic", "timeout", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|izd", kws,
```

Comment:
I don't think we want to expose the wonky C API here.
We should instead do:
`list_topics([topic=string], [timeout=None])`

if topic is set, use that, else return all topics.

So, just remove all_topics logic and pass it as `!topic`